File: uniconn-go/multi/multi_test.go
package multi_test

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net"
	"net/http"
	"net/http/httptest"
	"sync"
	"testing"
	"time"

	uniconn "github.com/snowmerak/uniconn/uniconn-go"
	"github.com/snowmerak/uniconn/uniconn-go/multi"
	"github.com/snowmerak/uniconn/uniconn-go/tcp"
	"github.com/snowmerak/uniconn/uniconn-go/websocket"
)

// ─── Negotiate ──────────────────────────────────────────────

func TestNegotiateHandler(t *testing.T) {
	entries := []multi.ProtocolEntry{
		{Name: multi.ProtoTCP, Address: "server:8001"},
		{Name: multi.ProtoWebSocket, Address: "ws://server:8002/ws"},
	}

	handler := multi.NegotiateHandler(entries, "")
	req := httptest.NewRequest(http.MethodGet, "/negotiate", nil)
	rec := httptest.NewRecorder()
	handler(rec, req)

	if rec.Code != http.StatusOK {
		t.Fatalf("status: %d", rec.Code)
	}
	if ct := rec.Header().Get("Content-Type"); ct != "application/json" {
		t.Fatalf("content-type: %s", ct)
	}

	var resp multi.NegotiateResponse
	if err := json.Unmarshal(rec.Body.Bytes(), &resp); err != nil {
		t.Fatalf("parse: %v", err)
	}
	if len(resp.Protocols) != 2 {
		t.Fatalf("got %d protocols, want 2", len(resp.Protocols))
	}
	if resp.Protocols[0].Name != multi.ProtoTCP {
		t.Fatalf("first protocol: %s", resp.Protocols[0].Name)
	}
	t.Logf("negotiate OK: %+v", resp)
}

func TestNegotiateHandlerRejectsPost(t *testing.T) {
	handler := multi.NegotiateHandler(nil, "")
	req := httptest.NewRequest(http.MethodPost, "/negotiate", nil)
	rec := httptest.NewRecorder()
	handler(rec, req)

	if rec.Code != http.StatusMethodNotAllowed {
		t.Fatalf("expected 405, got %d", rec.Code)
	}
}

// ─── MultiListener + MultiDialer Integration ────────────────

func TestMultiDialerSelectsBestProtocol(t *testing.T) {
	// Start a TCP listener.
	tcpLn, err := tcp.Listen("127.0.0.1:0", nil)
	if err != nil {
		t.Fatal(err)
	}

	// Start a WS listener.
	wsLn, err := websocket.Listen("127.0.0.1:0", &websocket.ListenConfig{Path: "/ws"})
	if err != nil {
		t.Fatal(err)
	}

	// Create MultiListener with negotiate endpoint.
	ml, err := multi.NewMultiListener("127.0.0.1:0",
		multi.TransportConfig{
			Protocol: multi.ProtoWebSocket,
			Address:  "ws://127.0.0.1:" + portFromAddr(wsLn.Addr()) + "/ws",
			Listener: tcpToUniconn(wsLn),
		},
		multi.TransportConfig{
			Protocol: multi.ProtoTCP,
			Address:  tcpLn.Addr().String(),
			Listener: tcpToUniconn(tcpLn),
		},
	)
	if err != nil {
		t.Fatal(err)
	}
	defer ml.Close()

	// Server: accept one connection and echo.
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		conn, proto, err := ml.AcceptWith()
		if err != nil {
			t.Errorf("accept: %v", err)
			return
		}
		t.Logf("accepted via %s", proto)
		defer conn.Close()
		io.Copy(conn, conn) // echo
	}()

	// Client: MultiDialer with only TCP dialer registered.
	// Should select TCP even though WS is higher priority,
	// because client has no WS dialer.
	negotiateURL := "http://" + ml.NegotiateAddr().String() + "/negotiate"

	md := multi.NewMultiDialer(multi.DialerConfig{
		NegotiateURL: negotiateURL,
		Dialers: map[multi.Protocol]uniconn.Dialer{
			multi.ProtoTCP: tcp.NewDialer(nil),
		},
		DialTimeout: 3 * time.Second,
	})

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	conn, proto, err := md.Dial(ctx)
	if err != nil {
		t.Fatalf("dial: %v", err)
	}
	defer conn.Close()

	if proto != multi.ProtoTCP {
		t.Fatalf("expected TCP, got %s", proto)
	}

	// Echo test.
	testMsg := []byte("hello multi-protocol!")
	if _, err := conn.Write(testMsg); err != nil {
		t.Fatalf("write: %v", err)
	}

	buf := make([]byte, 1024)
	n, err := conn.Read(buf)
	if err != nil {
		t.Fatalf("read: %v", err)
	}

	if !bytes.Equal(buf[:n], testMsg) {
		t.Fatalf("echo mismatch: got %q, want %q", buf[:n], testMsg)
	}

	t.Logf("multi-dialer echo via %s OK: %q", proto, testMsg)

	conn.Close()
	ml.Close()
	wg.Wait()
}

func TestMultiDialerFallback(t *testing.T) {
	// Start only a TCP listener.
	tcpLn, err := tcp.Listen("127.0.0.1:0", nil)
	if err != nil {
		t.Fatal(err)
	}

	// Negotiate advertises WebSocket AND TCP, but only TCP is actually listening.
	// Client has both dialers registered but WS address points to a closed port.
	closedPort := findFreePort(t)

	ml, err := multi.NewMultiListener("127.0.0.1:0",
		multi.TransportConfig{
			Protocol: multi.ProtoTCP,
			Address:  tcpLn.Addr().String(),
			Listener: tcpToUniconn(tcpLn),
		},
	)
	if err != nil {
		t.Fatal(err)
	}
	defer ml.Close()

	// server: echo
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		conn, _, err := ml.AcceptWith()
		if err != nil {
			return
		}
		defer conn.Close()
		io.Copy(conn, conn)
	}()

	// Manually build negotiate server with WS entry pointing to dead port.
	fakeNeg := httptest.NewServer(multi.NegotiateHandler([]multi.ProtocolEntry{
		{Name: multi.ProtoWebSocket, Address: "ws://127.0.0.1:" + closedPort + "/ws"},
		{Name: multi.ProtoTCP, Address: tcpLn.Addr().String()},
	}, ""))
	defer fakeNeg.Close()

	md := multi.NewMultiDialer(multi.DialerConfig{
		NegotiateURL: fakeNeg.URL + "/negotiate",
		Dialers: map[multi.Protocol]uniconn.Dialer{
			multi.ProtoWebSocket: websocket.NewDialer(nil),
			multi.ProtoTCP:       tcp.NewDialer(nil),
		},
		DialTimeout: 1 * time.Second,
	})

	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()

	conn, proto, err := md.Dial(ctx)
	if err != nil {
		t.Fatalf("dial: %v", err)
	}
	defer conn.Close()

	// Should have fallen back to TCP.
	if proto != multi.ProtoTCP {
		t.Fatalf("expected TCP fallback, got %s", proto)
	}

	t.Logf("fallback to %s OK", proto)

	conn.Close()
	ml.Close()
	wg.Wait()
}

// ─── Helpers ────────────────────────────────────────────────

// tcpToUniconn is a no-op adapter since our listeners already implement uniconn.Listener.
func tcpToUniconn(l uniconn.Listener) uniconn.Listener {
	return l
}

func portFromAddr(addr net.Addr) string {
	_, port, _ := net.SplitHostPort(addr.String())
	return port
}

func findFreePort(t *testing.T) string {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	port := portFromAddr(ln.Addr())
	ln.Close()
	return port
}

File: uniconn-js/node/src/multi/multi.test.ts
/**
 * Tests for multi-protocol negotiate, MultiDialer, and MultiListener.
 */

import { describe, it } from "node:test";
import * as assert from "node:assert/strict";
import * as http from "node:http";

import {
  MultiDialer,
  MultiListener,
  type ProtocolEntry,
  type NegotiateResponse,
  type TransportConfig,
  type Protocol,
} from "@uniconn/core/multi";
import type { IConn, IListener, IDialer, Addr, DialOptions } from "@uniconn/core/conn";

// ─── Mock implementations ───────────────────────────────────

class MockConn implements IConn {
  private data: Uint8Array[] = [];
  closed = false;

  async write(d: Uint8Array): Promise<number> {
    this.data.push(d.slice());
    return d.length;
  }
  async read(buffer: Uint8Array): Promise<number> {
    const d = this.data.shift();
    if (!d) return 0;
    buffer.set(d);
    return d.length;
  }
  async close() { this.closed = true; }
  localAddr(): Addr { return { network: "mock", address: "local" }; }
  remoteAddr(): Addr { return { network: "mock", address: "remote" }; }
  setDeadline() {}
  setReadDeadline() {}
  setWriteDeadline() {}
}

class MockListener implements IListener {
  private queue: IConn[] = [];
  private waiters: Array<(conn: IConn) => void> = [];
  private closed = false;

  push(conn: IConn) {
    const w = this.waiters.shift();
    if (w) { w(conn); } else { this.queue.push(conn); }
  }

  async accept(): Promise<IConn> {
    if (this.closed) throw new Error("closed");
    const q = this.queue.shift();
    if (q) return q;
    return new Promise<IConn>((r) => this.waiters.push(r));
  }
  async close() { this.closed = true; }
  addr(): Addr { return { network: "mock", address: "mock:0" }; }
}

class MockDialer implements IDialer {
  shouldFail: boolean;
  constructor(shouldFail = false) { this.shouldFail = shouldFail; }
  async dial(_addr: string, _opts?: DialOptions): Promise<IConn> {
    if (this.shouldFail) throw new Error("mock dial failure");
    return new MockConn();
  }
}

// ─── Tests ──────────────────────────────────────────────────

describe("NegotiateResponse", () => {
  it("serializes correctly", () => {
    const entries: ProtocolEntry[] = [
      { name: "tcp", address: "server:8001" },
      { name: "websocket", address: "ws://server:8002/ws" },
    ];
    const resp: NegotiateResponse = { protocols: entries };
    const json = JSON.stringify(resp);
    const parsed = JSON.parse(json) as NegotiateResponse;
    assert.equal(parsed.protocols.length, 2);
    assert.equal(parsed.protocols[0].name, "tcp");
  });
});

describe("MultiListener", () => {
  it("fans-in from multiple transports", async () => {
    const ln1 = new MockListener();
    const ln2 = new MockListener();

    const ml = new MultiListener([
      { protocol: "tcp" as Protocol, address: "host:1", listener: ln1 },
      { protocol: "websocket" as Protocol, address: "ws://host:2/ws", listener: ln2 },
    ]);

    // Push connections into both listeners.
    const conn1 = new MockConn();
    const conn2 = new MockConn();
    ln1.push(conn1);
    ln2.push(conn2);

    // Accept should get both.
    const r1 = await ml.acceptWith();
    const r2 = await ml.acceptWith();

    const protos = new Set([r1.protocol, r2.protocol]);
    assert.ok(protos.has("tcp"));
    assert.ok(protos.has("websocket"));

    await ml.close();
  });

  it("generates negotiate response", () => {
    const ml = new MultiListener([
      { protocol: "tcp" as Protocol, address: "host:1", listener: new MockListener() },
    ]);
    const neg = ml.getNegotiateResponse();
    assert.equal(neg.protocols.length, 1);
    assert.equal(neg.protocols[0].name, "tcp");
    ml.close();
  });
});

describe("MultiDialer", () => {
  it("selects best available protocol via negotiate", async () => {
    // Set up a mock negotiate HTTP server.
    const negResponse: NegotiateResponse = {
      protocols: [
        { name: "websocket", address: "ws://127.0.0.1:9999/ws" },
        { name: "tcp", address: "127.0.0.1:9998" },
      ],
    };

    const server = http.createServer((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify(negResponse));
    });

    await new Promise<void>((r) => server.listen(0, "127.0.0.1", r));
    const port = (server.address() as any).port;

    try {
      // Client only has TCP dialer → should select TCP.
      const md = new MultiDialer({
        negotiateURL: `http://127.0.0.1:${port}/negotiate`,
        dialers: { tcp: new MockDialer() },
        dialTimeout: 2000,
      });

      const { protocol } = await md.dial();
      assert.equal(protocol, "tcp");
    } finally {
      server.close();
    }
  });

  it("falls back when first protocol fails", async () => {
    const negResponse: NegotiateResponse = {
      protocols: [
        { name: "websocket", address: "ws://127.0.0.1:9999/ws" },
        { name: "tcp", address: "127.0.0.1:9998" },
      ],
    };

    const server = http.createServer((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify(negResponse));
    });

    await new Promise<void>((r) => server.listen(0, "127.0.0.1", r));
    const port = (server.address() as any).port;

    try {
      // WS dialer fails, TCP succeeds → should fall back to TCP.
      const md = new MultiDialer({
        negotiateURL: `http://127.0.0.1:${port}/negotiate`,
        dialers: {
          websocket: new MockDialer(true), // fails
          tcp: new MockDialer(false),      // succeeds
        },
        dialTimeout: 1000,
      });

      const { protocol } = await md.dial();
      assert.equal(protocol, "tcp");
    } finally {
      server.close();
    }
  });
});
